Document the advanced startup setup (#11915)
forem/forem: docs/getting-started/start-app.md

```

## Debugging the Rails application

[Overmind][overmind] lets you easily step through the Rails application in a
debugging session.

Using the `pry` gem, you can add a `binding.pry` to set a breakpoint in the method you're trying to
debug; the application will halt its execution there. You can then connect
to the web server by opening a separate terminal window and typing:

```shell
overmind c web
```

This will open up a [tmux][tmux] window pane at the debugging statement
position, which will look something like this:

```ruby
pry(#<Admin::AdminPortalsController>)> whereami

From: /forem/app/controllers/admin/admin_portals_controller.rb:8 Admin::AdminPortalsController#index:

    5: def index
    6:   a = "Hello debugger"
    7:   binding.pry
 => 8: end
```

## Inspecting the logs of each service

Overmind launches the various services required for our local setup: `web` (the
Rails web server), `webpacker` (the server managing JavaScript) and `sidekiq`
(the server managing the asynchronous queue).

If, for example, you want to inspect just the Sidekiq logs, you can open a
separate terminal window to look at those logs specifically:

```shell
overmind c sidekiq
```

> This will open a `tmux` console, which will allow you to browse _only_ the Sidekiq's logs.

Reviewer: ```suggestion
This will open a `tmux` console, which will allow you to browse _only_ the Sidekiq logs.
```

There are also some handy `tmux` shortcuts that you may find useful.

* The shortcut `C-b [` (_Control-b-open square bracket_) activates "scroll
mode", which allows you to use the arrows up and down and inspect the logs.

> * The shortcut `q` will deactivate "scroll mode".

Reviewer: ```suggestion
* The shortcut `q` deactivates "scroll mode".
```
Just to keep the tenses consistent :)

Current gotchas: potential environment issues with external services need to be
worked out.

# Starting the application (advanced)

> To have additional control on the local environment you might prefer using an

Reviewer: ```suggestion
To have additional control in your local environment, you might prefer using an
```

> advanced setup to start the application.

Reviewer: ```suggestion
advanced setup to start the application by using [Overmind][overmind].
```

The prerequisite is to install [Overmind][overmind], a process manager which
takes advantage of [tmux][tmux].

> This will allow you to launch everything in the same terminal, navigate the logs

Reviewer: ```suggestion
This will allow you to launch all your app's processes in the same terminal, navigate the logs
```
Not sure if this is actually correct, please use a better word if I am wrong in my suggestion here 😛 Just figured we could be a little more specific than "everything"

Author: definitely correct!

of each service separately, restart each service separately and have a better
debugging experience.

> After installing [Overmind][overmind] all you have to do to start the
> application is launch:

Reviewer: ```suggestion
After installing [Overmind][overmind], launch the application:
```

```shell
overmind s -f Procfile.dev
```

## Debugging the Rails application

[Overmind][overmind] lets you easily step through the Rails application in a
debugging session.

> If you use `binding.pry` to set a breakpoint in the method you're trying to
> debug the application will then halt its execution there. You can then connect
> to the web server by opening a separate terminal window and by typing:

Reviewer: ```suggestion
Using the `pry` gem, you can add a `binding.pry` to set a breakpoint in the method you're trying to
debug; the application will halt its execution there. You can then connect
to the web server by opening a separate terminal window and typing:
```

```shell
overmind c web
```

> this will open up a [tmux][tmux] window pane at the debugging statement

Reviewer: ```suggestion
This will open up a [tmux][tmux] window pane at the debugging statement
```

> position.
> 
> An example:

Reviewer: ```suggestion
position, which will look something like this:
```

```ruby
pry(#<Admin::AdminPortalsController>)> whereami

From: /forem/app/controllers/admin/admin_portals_controller.rb:8 Admin::AdminPortalsController#index:

    5: def index
    6:   a = "Hello debugger"
    7:   binding.pry
 => 8: end
```

## Inspecting the logs of each service

Overmind launches the various services required for our local setup: `web` (the
Rails web server), `webpacker` (the server managing JavaScript) and `sidekiq`
(the server managing the asynchronous queue).

> If, for example, you want to inspect the Sidekiq logs, proceed by opening a
> separate terminal window and typing:

Reviewer: ```suggestion
If, for example, you want to inspect just the Sidekiq logs, you can open a
separate terminal window to look at those logs specifically:
```

```shell
overmind c sidekiq
```

> this will open a `tmux` console allowing you to browse its logs.

Reviewer: ```suggestion
This will open a `tmux` console, which will allow you to browse _only_ the Sidekiq's logs.
```

> The shortcut `C-b [` (_Control-b-open square bracket_) will activate "scroll
> mode" allowing you to use the arrows up and down and inspect the logs, `q` will
> deactivate "scroll mode".

Reviewer: ```suggestion
There are also some handy `tmux` shortcuts that you may find useful.

* The shortcut `C-b [` (_Control-b-open square bracket_) activates "scroll
mode", which allows you to use the arrows up and down and inspect the logs.
* The shortcut `q` will deactivate "scroll mode".
```

> please refer to [tmux][tmux] documentation for configuration and for additional

Reviewer: ```suggestion
Please refer to [tmux][tmux] documentation for more information around `tmux` configuration and for additional
```

shortcuts.

## Resources

> Other than the official documentation on [Overmind][overmind] and [tmux][tmux]
> you can find the following resources useful:

Reviewer: ```suggestion
Other than the official [Overmind][overmind] and [tmux][tmux]
documentation, you may find the following resources useful:
```

- [Rails quick tips #6: tmux, tmuxinator and Overmind](https://dev.to/citizen428/rails-quick-tips-6-tmux-tmuxinator-and-overmind-4850)

> - [Give Your Terminal Super Powers: tmux Cheatsheet!](https://dev.to/jacobherrington/give-your-terminal-super-powers-tmux-cheatsheet-1p6p)

Reviewer: @citizen428 + @jacobherrington 👏🏽 🎉 💖
